web/src/stream_popover.js

```diff
@@ -214,7 +215,77 @@ function build_stream_popover(opts) {
     });
 }
 
-export function build_move_topic_to_stream_popover(
+async function get_message_placement_from_server(
+    current_stream_id,
+    topic_name,
+    current_message_id,
+) {
+    return new Promise((resolve) => {
+        message_edit.is_message_oldest_or_newest(
+            current_stream_id,
+            topic_name,
+            current_message_id,
+            (is_oldest, is_newest) => {
+                if (is_newest) {
+                    resolve("last");
+                } else if (is_oldest) {
+                    resolve("first");
+                } else {
+                    resolve("intermediate");
+                }
+            },
+        );
+    });
+}
+
+async function get_message_placement_in_conversation(
+    current_stream_id,
+    topic_name,
+    current_message_id,
+) {
```

Comment:
I tried to understand what supports_collapsing_recipients does, but do have a doubt over here. Does this verify if the messages are in a sorted order? Or does it verify if the message is in a group with consecutive other messages of the same topic so that they can be collapsed together?

If the later is true, then my logic of going through every message in the list is faulty, and it would have sufficed to only check the adjacent messages.

Reply:
It checks whether the view is one where if a message is the view, then all other messages in its conversation will be in the view. There's no guarantee they will be consecutive; "stream design" is an example such view.